README.md
[![Build Status](https://travis-ci.org/gambol99/keycloak-proxy.svg?branch=master)](https://travis-ci.org/gambol99/keycloak-proxy)
[![GoDoc](http://godoc.org/github.com/gambol99/keycloak-proxy?status.png)](http://godoc.org/github.com/gambol99/keycloak-proxy)

### **Keycloak Proxy**
----

  - Supports role based uri controls
  - Websocket connection upgrading
  - Token claim matching for additional ACL controls
  - Stateless offline refresh tokens with optional predefined session limits
  - TLS and mutual TLS support
  - JSON field bases access logs
  - Custom Sign-in and access forbidden pages
    
----

Keycloak-proxy is a proxy service which at the risk of stating the obvious integrates with the [Keycloak](https://github.com/keycloak/keycloak) authentication service. Although technically the service has no dependency on Keycloak itself and would quite happily work with any OpenID provider. The service supports both access tokens in browser cookie or bearer tokens.

```shell
[jest@starfury keycloak-proxy]$ bin/keycloak-proxy help
NAME:
   keycloak-proxy - is a proxy using the keycloak service for auth and authorization

USAGE:
   keycloak-proxy [global options] command [command options] [arguments...]
   
VERSION:
   v1.0.0-rc2
   
AUTHOR(S):
   Rohith <gambol99@gmail.com> 
   
COMMANDS:
   help, h	Shows a list of commands or help for one command
   
GLOBAL OPTIONS:
   --config 										        the path to the configuration file for the keycloak proxy
   --listen "127.0.0.1:3000"								the interface the service should be listening on
   --secret 										        the client secret used to authenticate to the oauth server
   --client-id 										        the client id used to authenticate to the oauth serves
   --discovery-url 									        the discovery url to retrieve the openid configuration
   --upstream-url "http://127.0.0.1:8081"					the url for the upstream endpoint you wish to proxy to
   --upstream-keepalives								    enables or disables the keepalive connections for upstream endpoint (defaults true)
   --encryption-key 									    the encryption key used to encrpytion the session state
   --redirection-url 									    the redirection url, namely the site url, note: /oauth will be added to it
   --hostname [--hostname option --hostname option]			a list of hostname which the service will respond to, defaults to all
   --tls-cert 										        the path to a certificate file used for TLS
   --tls-private-key 									    the path to the private key for TLS support
   --tls-ca-certificate 								    the path to the ca certificate used for mutual TLS
   --scope [--scope option --scope option]					a variable list of scopes requested when authenticating the user
   --claim [--claim option --claim option]					a series of key pair values which must match the claims in the token present e.g. aud=myapp, iss=http://example.com etcd
   --resource [--resource option --resource option]			a list of resources 'uri=/admin|methods=GET|roles=role1,role2'
   --signin-page 									        a custom template displayed for signin
   --forbidden-page 									    a custom template used for access forbidden
   --tag [--tag option --tag option]						a keypair tag which is passed to the templates when render, i.e. title='My Page',site='my name' etc
   --max-session "1h0m0s"							      	if refresh sessions are enabled we can limit their duration via this
   --cors-origins [--cors-origins option --cors-origins option]				a set of origins to add to the CORS access control (Access-Control-Allow-Origin)
   --cors-methods [--cors-methods option --cors-methods option]				the method permitted in the access control (Access-Control-Allow-Methods)
   --cors-headers [--cors-headers option --cors-headers option]				a set of headers to add to the CORS access control (Access-Control-Allow-Headers)
   --cors-exposes-headers [--cors-exposes-headers option --cors-exposes-headers option]	set the expose cors headers access control (Access-Control-Expose-Headers)
   --cors-max-age "0"									    the max age applied to cors headers (Access-Control-Max-Age)
   --cors-credentials									    the credentials access control header (Access-Control-Allow-Credentials)
   --skip-token-verification								testing purposes ONLY, the option allows you to bypass the token verification, expiration and roles are still enforced
   --proxy-protocol									        switches on proxy protocol support on the listen (not supported yet)
   --refresh-sessions									    enables the refreshing of tokens via offline access (defaults false)
   --json-logging									        switch on json logging rather than text (defaults true)
   --log-requests									        switch on logging of all incoming requests (defaults true)
   --verbose										        switch on debug / verbose logging
   --help, -h										        show help
   --version, -v									        print the version

```

#### **Configuration**

The configuration can come from a yaml/json file and or the command line options (note, command options have a higher priority and will override any options referenced in a config file)

```YAML
# is the url for retrieve the openid configuration - normally the <server>/auth/realm/<realm_name>
discovery_url: https://keycloak.example.com/auth/realms/<REALM_NAME>
# the client id for the 'client' application
clientid: <CLIENT_ID>
# the secret associated to the 'client' application
secret: <CLIENT_SECRET>
# the interface definition you wish the proxy to listen, all interfaces is specified as ':<port>'
listen: 127.0.0.1:3000
# whether to request offline access and use a refresh token
refresh_session: true
# assuming you are using refresh tokens, specify the maximum amount of time the refresh token can last
max_session: 1h
# the location of a certificate you wish the proxy to use for TLS support
tls_cert:
# the location of a private key for TLS
tls_private_key:
# the redirection url, essentially the site url, note: /oauth/callback is added at the end
redirection_url: http://127.0.0.3000
# the encryption key used to encode the session state
encryption_key: <ENCRYPTION_KEY>
# the upstream endpoint which we should proxy request
upstream: http://127.0.0.1:80
# additional scopes to add to add to the default (openid+email+profile)
scopes:
  - vpn-user

# a collection of resource i.e. urls that you wish to protect
resources:
  - url: /admin/test
    # the methods on this url that should be protected, if missing, we assuming all
    methods:
      - GET
    # a list of roles the user must have in order to accces urls under the above
    roles:
      - openvpn:vpn-user
      - openvpn:prod-vpn
      - test
  - url: /admin
    methods:
      - GET
    roles:
      - openvpn:vpn-user
      - openvpn:commons-prod-vpn
```

#### **Example Usage**

Assuming you have some web service you wish protected by Keycloak;

a) Create the *client* under the Keycloak GUI or CLI; the client protocol is *'openid-connect'*, access-type:  *confidential*.
b) Add a Valid Redirect URIs of *http://127.0.0.1:3000/oauth/callback*.
c) Grab the client id and client secret.
d) Create the various roles under the client or existing clients for authorization purposes.

**The default config**

```YAML
discovery_url: https://keycloak.example.com/auth/realms/<REALM_NAME>
clientid: <CLIENT_ID>
secret: <CLIENT_SECRET>
listen: 127.0.0.1:3000
redirection_url: http://127.0.0.3000
refresh_session: false
encryption_key: AgXa7xRcoClDEU0ZDSH4X0XhL5Qy2Z2j
upstream: http://127.0.0.1:80

resources:
  - url: /admin
    methods:
      - GET
    roles:
      - client:test1
      - client:test2
  - url: /backend
    roles:
      - client:test1
```

Note, anything defined in the configuration file can also be configured as command line options, so the above would be reflected as;

```shell
bin/keycloak-proxy \
    --discovery-url=https://keycloak.example.com/auth/realms/<REALM_NAME> \
    --client-id=<CLIENT_ID> \
    --secret=<SECRET> \
    --listen=127.0.0.1:3000 \
    --redirection-url=http://127.0.0.3000 \
    --refresh-sessions=true \
    --encryption-key=AgXa7xRcoClDEU0ZDSH4X0XhL5Qy2Z2j \
    --upstream=http://127.0.0.1:80 \
    --resource="uri=/admin|methods=GET|roles=test1,test2" \
    --resource="uri=/backend|roles=test1"
```

#### **Google OpenID**

Although the role extensions do require a Keycloak IDP or at the very least a IDP that produces a token which contains roles, there's nothing stopping you from using it against any OpenID providers, such as Google. Go to the Google Developers Console and create a new application *(via "Enable and Manage APIs -> Credentials)*. Once you've created the application, take the client id, secret and make sure you've added the callback url to the application scope *(using the default this would be http://127.0.0.1:3000/oauth/callback)* 

``` shell
bin/keycloak-proxy \
    --discovery-url=https://accounts.google.com/.well-known/openid-configuration \
    --client-id=<CLIENT_ID> \
    --secret=<CLIENT_SECRET> \
    --resource="uri=/" \   
    --verbose=true
```

Open a browser an go to http://127.0.0.1:3000 and you should be redirected to Google for authenticate and back the application when done and you should see something like the below.

```shell
DEBU[0002] resource access permitted: /                  access=permitted bearer=false expires=57m51.32029042s resource=/ username=gambol99@gmail.com
2016-02-06 13:59:01.680300 I | http: proxy error: dial tcp 127.0.0.1:8081: getsockopt: connection refused
DEBU[0002] resource access permitted: /favicon.ico       access=permitted bearer=false expires=57m51.144004098s resource=/ username=gambol99@gmail.com
2016-02-06 13:59:01.856716 I | http: proxy error: dial tcp 127.0.0.1:8081: getsockopt: connection refused
```

#### **Upstream Headers**

On protected resources the upstream endpoint will receive a number of headers added by the proxy;

```GO
# add the header to the upstream endpoint
cx.Request.Header.Add("X-Auth-UserId", id.id)
cx.Request.Header.Add("X-Auth-Subject", id.preferredName)
cx.Request.Header.Add("X-Auth-Username", id.name)
cx.Request.Header.Add("X-Auth-Email", id.email)
cx.Request.Header.Add("X-Auth-ExpiresIn", id.expiresAt.String())
cx.Request.Header.Add("X-Auth-Token", id.token.Encode())
cx.Request.Header.Add("X-Auth-Roles", strings.Join(id.roles, ","))

# plus the default
cx.Request.Header.Add("X-Forwarded-For", <CLIENT_IP>)
cx.Request.Header.Add("X-Forwarded-Proto", <CLIENT_PROTO>)
```

#### **Encryption Key**

In order to remain stateless and not have to rely on a central cache to persist the 'refresh_tokens', the refresh token is encrypted and added as a cookie using *crypto/aes*. 
Naturally the key must be the same if your running behind a load balancer etc.  

#### **Claim Matching**

The proxy supports adding a variable list of claim matches against the presented tokens for additional access control. So for example you can match the 'iss' or 'aud' to the token or custom attributes; 
note each of the matches are regex's. Examples,  --claim 'aud=sso.*' --claim iss=https://.*'

#### **Custom Pages**

By default the proxy will immediately redirect you for authentication and hand back 403 for access denied. Most users will probably want to present the user with a more friendly
sign-in and access denied page. You can pass the command line options (or via config file) paths to the files i.e. --signin-pag=PATH. The sign-in page will have a 'redirect' 
passed into the scope hold the oauth redirection url. If you wish pass additional variables into the templates, perhaps title, sitename etc, you can use the --tag key=pair i.e. 
--tag title="This is my site"; the variable would be accessible from {{ .title }}

```HTML
<html>
<body>
<a href="{{ .redirect }}">Sign-in</a>
</body>
</html>
```

#### **White-listed URL's** 

Depending on how the application urls are laid out, you might want protect the root / url but have exceptions on a list of paths, i.e. /health etc. Although you should probably
fix this by fixing up the paths, you can add excepts to the protected resources. (Note: it's an array, so the order is important)

```YAML
  resources:
  - url: /some_white_listed_url
    white-listed: true
  - url: /
    methods:
      - GET
    roles:
      - <CLIENT_APP_NAME>:<ROLE_NAME>
      - <CLIENT_APP_NAME>:<ROLE_NAME>
```

Or on the command line 

```shell
  --resource "uri=/some_white_listed_url,white-listed=true"
  --resource "uri=/"  # requires authentication on the rest
  --resource "uri=/admin|roles=admin,superuser|methods=POST,DELETE
```

#### **Mutual TLS**

The proxy support enforcing mutual TLS for the clients by simply adding the --tls-ca-certificate command line option or config file option. All clients connecting must present a ceritificate
which was signed by the CA being used.

#### **Refresh Tokens**

Refresh tokens (--refresh-sessions=true) request a offline access type from the provider and will refresh the access token as and when they expire. By default the refresh tokens are dropped in as
an encrypted cookie (--encryption-key=KEY). Alternatively you can place the refresh token (still requires encryption key) in a local boltdb file or shared redis. Naturally the encryption key has to 
be the same on all instances and boltdb is for single instance only developments.

#### **Endpoints**

* **/oauth/authorize** is authentication endpoint which will generate the openid redirect to the provider
* **/oauth/callback** is provider openid callback endpoint
* **/oauth/expired** is a helper endpoint to check if a access token has expired, 200 for ok and, 401 for no token and 401 for expired
* **/oauth/token** is a helper endpoint which will display the current access token for you
* **/oauth/health** is the health checking endpoint for the proxy



CHANGELOG.md

#### **1.0.2 (April 15th, 2016)**

FIXES:
 * Cleaned up a lot of code base
 * Fixed elements in the refresh tokens and simplified the controller
 * Removed of the code out from methods into functions to reduce the dependencies (unit testing is easier as well)

FEATURES:
 * Added the ability to store the refresh token in either local boltdb file or a redis service rather than
   an encrypted cookie (note, the token regardless is encrypted)
 * Added prometheus metrics endpoints /oauth/metrics

BREAKING CHANGES:
 * Changed the following configuration options to conform to their command line equivalents
   - refresh_sessions -> refresh-sessions
   - discovery_url      -> discovery-url
   - redirection_url    -> redirection-url
   - tls_ca_certificate -> tls-ca-certificate
   - tls_private_key    -> tls-private-key
   - tls_cert           -> tls-cert
   - log_json_format    -> log-json-format
   - log_requests       -> log-requests
   - forbidden_page     -> forbidden-page
   - sign_in_page       -> sign-in-page

#### **1.0.1 (April 8th, 2016)**

FIXES:
 * Fixed the refresh tokens for those provides whom do not use JWT tokens, Google Connect for example

#### **1.0.0 (April 8th, 2016)**

FEATURES
 * Added the /oauth/expiration controller to test for access token expiration
 * Added the /oauth/token as a helper method to display the access token

FIXES:
 * Fixed and cleaned up a few niggling issues

#### **1.0.0-rc6 (March 31th, 2016)**

FIXES:
 * Added a option to control the upstream TLS verification
 * Added in the x-forwarded-for headers rather than overwriting
 * Moved back to using the official coreos go-oidc rather than the hacked version

#### **1.0.0-rc5 (March 15th, 2016)**

FEATURES:

 * Added the realm access roles for keycloak, beforehand the user contect
   was only parses roles which were from client applications

BUGS:

 * Fixed the gitlab-ci build scripts
 * Fixed the custom forbidden page bug


config.go
/*
Copyright 2015 All rights reserved.
Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package main

import (
	"encoding/json"
	"fmt"
	"io/ioutil"
	"net/url"
	"path/filepath"
	"regexp"
	"strings"
	"time"

	"gopkg.in/yaml.v2"

	"github.com/codegangsta/cli"
)

// newDefaultConfig returns a initialized config
func newDefaultConfig() *Config {
	return &Config{
		Listen:         "127.0.0.1:3000",
		RedirectionURL: "http://127.0.0.1:3000",
		Upstream:       "http://127.0.0.1:8081",
		MaxSession:     time.Duration(1) * time.Hour,
		TagData:        make(map[string]string, 0),
		ClaimsMatch:    make(map[string]string, 0),
		Header:         make(map[string]string, 0),
		CORS:           &CORS{},
		SkipUpstreamTLSVerify: true,
	}
}

// isValid validates if the config is valid
func (r *Config) isValid() error {
	if r.Upstream == "" {
		return fmt.Errorf("you have not specified an upstream endpoint to proxy to")
	}
	if _, err := url.Parse(r.Upstream); err != nil {
		return fmt.Errorf("the upstream endpoint is invalid, %s", err)
	}
	if r.Listen == "" {
		return fmt.Errorf("you have not specified the listening interface")
	}
	if r.TLSCertificate != "" && r.TLSPrivateKey == "" {
		return fmt.Errorf("you have not provided a private key")
	}
	if r.TLSPrivateKey != "" && r.TLSCertificate == "" {
		return fmt.Errorf("you have not provided a certificate file")
	}
	if r.TLSCertificate != "" && !fileExists(r.TLSCertificate) {
		return fmt.Errorf("the tls certificate %s does not exist", r.TLSCertificate)
	}
	if r.TLSPrivateKey != "" && !fileExists(r.TLSPrivateKey) {
		return fmt.Errorf("the tls private key %s does not exist", r.TLSPrivateKey)
	}
	if r.TLSCaCertificate != "" && !fileExists(r.TLSCaCertificate) {
		return fmt.Errorf("the tls ca certificate file %s does not exist", r.TLSCaCertificate)
	}
	// step: if the skip verification is off, we need the below
	if !r.SkipTokenVerification {
		if r.DiscoveryURL == "" {
			return fmt.Errorf("you have not specified the discovery url")
		}
		if r.ClientID == "" {
			return fmt.Errorf("you have not specified the client id")
		}
		if r.Secret == "" {
			return fmt.Errorf("you have not specified the client secret")
		}
		if r.RedirectionURL == "" {
			return fmt.Errorf("you have not specified the redirection url")
		}
		if strings.HasSuffix(r.RedirectionURL, "/") {
			r.RedirectionURL = strings.TrimSuffix(r.RedirectionURL, "/")
		}
		if r.EncryptionKey == "" && r.RefreshSessions {
			return fmt.Errorf("you have not specified a encryption key for encoding the session state")
		}
		if r.EncryptionKey != "" && len(r.EncryptionKey) < 32 {
			return fmt.Errorf("the encryption key is too short, must be longer than 32 characters")
		}
		if r.StoreURL != "" {
			if _, err := url.Parse(r.StoreURL); err != nil {
				return fmt.Errorf("the store url is invalid, error: %s", err)
			}
		}
		if r.MaxSession == 0 && r.RefreshSessions {
			r.MaxSession = time.Duration(6) * time.Hour
		}
	}
	// step: valid the resources
	for _, resource := range r.Resources {
		if err := resource.IsValid(); err != nil {
			return err
		}
	}
	// step: validate the claims are validate regex's
	for k, claim := range r.ClaimsMatch {
		// step: validate the regex
		if _, err := regexp.Compile(claim); err != nil {
			return fmt.Errorf("the claim matcher: %s for claim: %s is not a valid regex", claim, k)
		}
	}

	return nil
}

// hasSignInPage checks if there is a custom sign in  page
func (r *Config) hasSignInPage() bool {
	if r.SignInPage != "" {
		return true
	}

	return false
}

// hasForbiddenPage checks if there is a custom forbidden page
func (r *Config) hasForbiddenPage() bool {
	if r.ForbiddenPage != "" {
		return true
	}

	return false
}

// readOptions parses the command line options and constructs a config object
func readOptions(cx *cli.Context, config *Config) (err error) {
	if cx.IsSet("listen") {
		config.Listen = cx.String("listen")
	}
	if cx.IsSet("secret") {
		config.Secret = cx.String("secret")
	}
	if cx.IsSet("client-id") {
		config.ClientID = cx.String("client-id")
	}
	if cx.IsSet("discovery-url") {
		config.DiscoveryURL = cx.String("discovery-url")
	}
	if cx.IsSet("upstream-url") {
		config.Upstream = cx.String("upstream-url")
	}
	if cx.IsSet("upstream-keepalives") {
		config.Keepalives = cx.Bool("upstream-keepalives")
	}
	if cx.IsSet("skip-token-verification") {
		config.SkipTokenVerification = cx.Bool("skip-token-verification")
	}
	if cx.IsSet("skip-upstream-tls-verify") {
		config.SkipUpstreamTLSVerify = cx.Bool("skip-upstream-tls-verify")
	}
	if cx.IsSet("encryption-key") {
		config.EncryptionKey = cx.String("encryption-key")
	}
	if cx.IsSet("store-url") {
		config.StoreURL = cx.String("store-url")
	}
	if cx.IsSet("no-redirects") {
		config.NoRedirects = cx.Bool("no-redirects")
	}
	if cx.IsSet("redirection-url") {
		config.RedirectionURL = cx.String("redirection-url")
	}
	if cx.IsSet("tls-cert") {
		config.TLSCertificate = cx.String("tls-cert")
	}
	if cx.IsSet("tls-private-key") {
		config.TLSPrivateKey = cx.String("tls-private-key")
	}
	if cx.IsSet("tls-ca-certificate") {
		config.TLSCaCertificate = cx.String("tls-ca-certificate")
	}
	if cx.IsSet("signin-page") {
		config.SignInPage = cx.String("signin-page")
	}
	if cx.IsSet("forbidden-page") {
		config.ForbiddenPage = cx.String("forbidden-page")
	}
	if cx.IsSet("max-session") {
		config.MaxSession = cx.Duration("max-session")
	}
	if cx.IsSet("enable-security-filter") {
		config.EnableSecurityFilter = true
	}
	if cx.IsSet("proxy-protocol") {
		config.ProxyProtocol = cx.Bool("proxy-protocol")
	}
	if cx.IsSet("refresh-sessions") {
		config.RefreshSessions = cx.Bool("refresh-sessions")
	}
	if cx.IsSet("json-logging") {
		config.LogJSONFormat = cx.Bool("json-logging")
	}
	if cx.IsSet("log-requests") {
		config.LogRequests = cx.Bool("log-requests")
	}
	if cx.IsSet("verbose") {
		config.Verbose = cx.Bool("verbose")
	}
	if cx.IsSet("scope") {
		config.Scopes = cx.StringSlice("scope")
	}
	if cx.IsSet("hostname") {
		config.Hostnames = cx.StringSlice("hostname")
	}
	if cx.IsSet("cors-origins") {
		config.CORS.Origins = cx.StringSlice("cors-origins")
	}
	if cx.IsSet("cors-methods") {
		config.CORS.Methods = cx.StringSlice("cors-methods")
	}
	if cx.IsSet("cors-headers") {
		config.CORS.Headers = cx.StringSlice("cors-headers")
	}
	if cx.IsSet("cors-exposed-headers") {
		config.CORS.ExposedHeaders = cx.StringSlice("cors-exposed-headers")
	}
	if cx.IsSet("cors-max-age") {
		config.CORS.MaxAge = cx.Duration("cors-max-age")
	}
	if cx.IsSet("cors-credentials") {
		config.CORS.Credentials = cx.BoolT("cors-credentials")
	}
	if cx.IsSet("tag") {
		config.TagData, err = decodeKeyPairs(cx.StringSlice("tag"))
		if err != nil {
			return err
		}
	}
	if cx.IsSet("claim") {
		config.ClaimsMatch, err = decodeKeyPairs(cx.StringSlice("claim"))
		if err != nil {
			return err
		}
	}
	if cx.IsSet("header") {
		config.Header, err = decodeKeyPairs(cx.StringSlice("header"))
		if err != nil {
			return err
		}
	}
	if cx.IsSet("resource") {
		for _, x := range cx.StringSlice("resource") {
			resource, err := decodeResource(x)
			if err != nil {
				return fmt.Errorf("invalid resource %s, %s", x, err)
			}
			config.Resources = append(config.Resources, resource)
		}
	}

	return nil
}

// readConfigFile reads and parses the configuration file
func readConfigFile(filename string, config *Config) error {
	// step: read in the contents of the file
	content, err := ioutil.ReadFile(filename)
	if err != nil {
		return err
	}
	// step: attempt to un-marshal the data
	switch ext := filepath.Ext(filename); ext {
	case "json":
		err = json.Unmarshal(content, config)
	default:
		err = yaml.Unmarshal(content, config)
	}

	return err
}

// getOptions returns the command line options
func getOptions() []cli.Flag {
	defaults := newDefaultConfig()

	return []cli.Flag{
		cli.StringFlag{
			Name:  "config",
			Usage: "the path to the configuration file for the keycloak proxy",
		},
		cli.StringFlag{
			Name:  "listen",
			Usage: "the interface the service should be listening on",
			Value: defaults.Listen,
		},
		cli.StringFlag{
			Name:  "secret",
			Usage: "the client secret used to authenticate to the oauth server",
		},
		cli.StringFlag{
			Name:  "client-id",
			Usage: "the client id used to authenticate to the oauth serves",
		},
		cli.StringFlag{
			Name:  "discovery-url",
			Usage: "the discovery url to retrieve the openid configuration",
		},
		cli.StringFlag{
			Name:  "upstream-url",
			Usage: "the url for the upstream endpoint you wish to proxy to",
			Value: defaults.Upstream,
		},
		cli.BoolTFlag{
			Name:  "upstream-keepalives",
			Usage: "enables or disables the keepalive connections for upstream endpoint (defaults true)",
		},
		cli.StringFlag{
			Name:  "encryption-key",
			Usage: "the encryption key used to encrpytion the session state",
		},
		cli.StringFlag{
			Name:  "store-url",
			Usage: "the store url to use for storing the refresh tokens, i.e. redis://127.0.0.1:6379, file:///etc/tokens.file",
		},
		cli.BoolFlag{
			Name:  "no-redirects",
			Usage: "do not have back redirects when no authentication is present, simple reply with 401 code",
		},
		cli.StringFlag{
			Name:  "redirection-url",
			Usage: fmt.Sprintf("the redirection url, namely the site url, note: %s will be added to it", oauthURL),
		},
		cli.StringSliceFlag{
			Name:  "hostname",
			Usage: "a list of hostnames the service will respond to, defaults to all",
		},
		cli.StringFlag{
			Name:  "tls-cert",
			Usage: "the path to a certificate file used for TLS",
		},
		cli.StringFlag{
			Name:  "tls-private-key",
			Usage: "the path to the private key for TLS support",
		},
		cli.StringFlag{
			Name:  "tls-ca-certificate",
			Usage: "the path to the ca certificate used for mutual TLS",
		},
		cli.BoolTFlag{
			Name:  "skip-upstream-tls-verify",
			Usage: "whether to skip the verification of any upstream TLS (defaults to true)",
		},
		cli.StringSliceFlag{
			Name:  "scope",
			Usage: "a variable list of scopes requested when authenticating the user",
		},
		cli.StringSliceFlag{
			Name:  "claim",
			Usage: "a series of key pair values which must match the claims in the token present e.g. aud=myapp, iss=http://example.com etcd",
		},
		cli.StringSliceFlag{
			Name:  "resource",
			Usage: "a list of resources 'uri=/admin|methods=GET|roles=role1,role2'",
		},
		cli.StringFlag{
			Name:  "signin-page",
			Usage: "a custom template displayed for signin",
		},
		cli.StringFlag{
			Name:  "forbidden-page",
			Usage: "a custom template used for access forbidden",
		},
		cli.StringSliceFlag{
			Name:  "tag",
			Usage: "a keypair tag which is passed to the templates when render, i.e. title='My Page',site='my name' etc",
		},
		cli.DurationFlag{
			Name:  "max-session",
			Usage: "if refresh sessions are enabled we can limit their duration via this",
			Value: defaults.MaxSession,
		},
		cli.StringSliceFlag{
			Name:  "cors-origins",
			Usage: "a set of origins to add to the CORS access control (Access-Control-Allow-Origin)",
		},
		cli.StringSliceFlag{
			Name:  "cors-methods",
			Usage: "the method permitted in the access control (Access-Control-Allow-Methods)",
		},
		cli.StringSliceFlag{
			Name:  "cors-headers",
			Usage: "a set of headers to add to the CORS access control (Access-Control-Allow-Headers)",
		},
		cli.StringSliceFlag{
			Name:  "cors-exposes-headers",
			Usage: "set the expose cors headers access control (Access-Control-Expose-Headers)",
		},
		cli.DurationFlag{
			Name:  "cors-max-age",
			Usage: "the max age applied to cors headers (Access-Control-Max-Age)",
		},
		cli.BoolFlag{
			Name:  "cors-credentials",
			Usage: "the credentials access control header (Access-Control-Allow-Credentials)",
		},
		cli.BoolFlag{
			Name:  "enable-security-filter",
			Usage: "enables the security filter handler",
		},
		cli.BoolFlag{
			Name:  "skip-token-verification",
			Usage: "testing purposes ONLY, the option allows you to bypass the token verification, expiration and roles are still enforced",
		},
		cli.BoolFlag{
			Name:  "proxy-protocol",
			Usage: "switches on proxy protocol support on the listen (not supported yet)",
		},
		cli.BoolFlag{
			Name:  "refresh-sessions",
			Usage: "enables the refreshing of tokens via offline access (defaults false)",
		},
		cli.BoolTFlag{
			Name:  "json-logging",
			Usage: "switch on json logging rather than text (defaults true)",
		},
		cli.BoolTFlag{
			Name:  "log-requests",
			Usage: "switch on logging of all incoming requests (defaults true)",
		},
		cli.BoolFlag{
			Name:  "verbose",
			Usage: "switch on debug / verbose logging",
		},
	}
}
